crates/oxc_linter/src/rules/unicorn/number_literal_case.rs

```diff
@@ -0,0 +1,246 @@
+    if number_literal.starts_with("0B") || number_literal.starts_with("0O") {
+        return Some((
+            Span { start: raw_span.start + 1, end: raw_span.start + 2 },
+            Message::UppercasePrefix,
+            number_literal.to_lowercase(),
+        ));
+    }
+    if number_literal.starts_with("0X") || number_literal.starts_with("0x") {
+        let has_uppercase_prefix = number_literal.starts_with("0X");
+        let has_lowercase_digits = number_literal[2..].chars().any(|c| ('a'..='f').contains(&c));
+        if has_uppercase_prefix && has_lowercase_digits {
+            return Some((
+                raw_span,
+                Message::UppercasePrefixAndLowercaseHexadecimalDigits,
+                "0x".to_owned() + &digits_to_uppercase(&number_literal[2..]),
+            ));
+        }
+        if has_uppercase_prefix {
+            return Some((
+                Span { start: raw_span.start + 1, end: raw_span.start + 2 },
+                Message::UppercasePrefix,
+                "0x".to_owned() + &number_literal[2..],
+            ));
+        }
+        if has_lowercase_digits {
+            return Some((
+                Span { start: raw_span.start + 2, end: raw_span.end },
+                Message::LowercaseHexadecimalDigits,
+                "0x".to_owned() + &digits_to_uppercase(&number_literal[2..]),
+            ));
+        }
+        return None;
+    }
+    if let Some(index) = number_literal.find('E') {
+        let char_position = raw_span.start + index as u32;
+        return Some((
+            Span { start: char_position, end: char_position + 1 },
+            Message::UppercaseExponentialNotation,
+            number_literal.to_lowercase(),
+        ));
+    }
+    None
+}
+
+fn digits_to_uppercase(digits: &str) -> String {
+    let mut result = digits.to_uppercase();
+    if result.ends_with('N') {
+        result.truncate(result.len() - 1);
+        result.push('n');
+    }
+    result
+}
+
+enum Message {
+    UppercasePrefix,
+    UppercaseExponentialNotation,
+    LowercaseHexadecimalDigits,
+    UppercasePrefixAndLowercaseHexadecimalDigits,
+}
+
+impl Message {
+    fn details(&self) -> (&'static str, &'static str) {
+        match self {
+            Self::UppercasePrefix => ("Unexpected number literal prefix in uppercase.", "Use lowercase for number literal prefixes (`0b`, `0o`, and `0x`)."),
+            Self::UppercaseExponentialNotation => ("Unexpected exponential notation in uppercase.", "Use lowercase for `e` in exponential notations."),
+            Self::LowercaseHexadecimalDigits => ("Unexpected hexadecimal digits in lowercase.", "Use uppercase for hexadecimal digits."),
+            Self::UppercasePrefixAndLowercaseHexadecimalDigits => ("Unexpected number literal prefix in uppercase and hexadecimal digits in lowercase.", "Use lowercase for number literal prefixes (`0b`, `0o`, and `0x`) and uppercase for hexadecimal digits.")
+        }
+    }
+}
```

Comment:
shall we make the diagnostic an enum instead?


https://github.com/oxc-project/oxc/blob/db5417f9a954b18a7dfd867b1ee9d7d7b7655db2/crates/oxc_linter/src/rules/eslint/valid_typeof.rs#L13-L21

Reply:
Thanks for the suggestion! Updated in 2a58b6b77859ad28891c7036d007b834f93dee5f. I slightly improved the help message for some cases.

BTW, I also changed the category of this rule from "correctness" to "style" in 2632f65ce93f6f8fe8b2acdc1c61d0ae615a87dc.